File: sdk/typescript/src/crypto.ts
import { createHash } from 'crypto';

// Types
export type HashFn = (left: bigint, right: bigint) => bigint;

export interface MerkleProof {
  path: bigint[];
  indices: number[];
  root: bigint;
  leaf: bigint;
}

/**
 * Minimal binary Merkle tree. Default hash is Poseidon if available; otherwise it will
 * throw when trying to hash unless a custom hashFn is provided.
 */
export class MerkleTree {
  private leaves: bigint[];
  private layers: bigint[][];
  private hashFn: HashFn;
  private zero: bigint;

  constructor(
    leaves: bigint[],
    options: {
      hashFn?: HashFn;
      depth?: number;
      zero?: bigint;
    } = {}
  ) {
    this.zero = options.zero ?? 0n;
    this.hashFn = options.hashFn ?? defaultPoseidonHashFn();
    const depth = options.depth ?? Math.ceil(Math.log2(Math.max(2, leaves.length))) + 1;
    this.leaves = this.padLeaves(leaves, depth);
    this.layers = this.buildLayers(this.leaves);
  }

  static async createWithPoseidon(
    leaves: bigint[],
    options: { depth?: number; zero?: bigint } = {}
  ): Promise<MerkleTree> {
    const hashFn = await createPoseidonHashFn();
    return new MerkleTree(leaves, { ...options, hashFn });
  }

  getRoot(): bigint {
    return this.layers[this.layers.length - 1][0];
  }

  getProof(index: number): MerkleProof {
    if (index < 0 || index >= this.leaves.length) {
      throw new Error('Leaf index out of range');
    }
    const path: bigint[] = [];
    const indices: number[] = [];
    let idx = index;
    for (let level = 0; level < this.layers.length - 1; level += 1) {
      const isRight = idx % 2;
      const siblingIdx = isRight ? idx - 1 : idx + 1;
      const sibling = this.layers[level][siblingIdx] ?? this.zero;
      path.push(sibling);
      indices.push(isRight);
      idx = Math.floor(idx / 2);
    }

    return { path, indices, root: this.getRoot(), leaf: this.leaves[index] };
  }

  verify(leaf: bigint, proof: MerkleProof): boolean {
    return MerkleTree.verifyProof(leaf, proof, this.hashFn);
  }

  static verifyProof(leaf: bigint, proof: MerkleProof, hashFn?: HashFn): boolean {
    const hf = hashFn ?? defaultPoseidonHashFn();
    let acc = leaf;
    for (let i = 0; i < proof.path.length; i += 1) {
      const sibling = proof.path[i];
      const isRight = proof.indices[i] === 1;
      acc = isRight ? hf(sibling, acc) : hf(acc, sibling);
    }
    return acc === proof.root;
  }

  private padLeaves(leaves: bigint[], depth: number): bigint[] {
    const size = 2 ** (depth - 1);
    if (leaves.length > size) {
      throw new Error(`Too many leaves for depth ${depth}: got ${leaves.length}, max ${size}`);
    }
    const padded = [...leaves];
    while (padded.length < size) padded.push(this.zero);
    return padded;
  }

  private buildLayers(leaves: bigint[]): bigint[][] {
    const layers: bigint[][] = [leaves];
    let current = leaves;
    while (current.length > 1) {
      const next: bigint[] = [];
      for (let i = 0; i < current.length; i += 2) {
        const left = current[i];
        const right = current[i + 1] ?? this.zero;
        next.push(this.hashFn(left, right));
      }
      layers.push(next);
      current = next;
    }
    return layers;
  }
}

// Poseidon utilities (lazy loaded)
let poseidonCache: any | null | undefined; // undefined: not tried, null: unavailable

export async function loadPoseidon(): Promise<any> {
  if (poseidonCache !== undefined) {
    if (poseidonCache === null) {
      throw new Error('circomlibjs (poseidon) is not available. Install it with `bun add circomlibjs`.');
    }
    return poseidonCache;
  }
  try {
    const mod = await import('circomlibjs');
    const builder = (mod as any).buildPoseidon ?? (mod as any).poseidon ?? (mod as any).buildPoseidonOpt;
    if (!builder) throw new Error('Poseidon builder not found in circomlibjs');
    const poseidon = await builder();
    poseidonCache = poseidon;
    return poseidon;
  } catch (err) {
    poseidonCache = null;
    throw new Error(
      'circomlibjs is required for Poseidon operations. Install it with `bun add circomlibjs` in your project.'
    );
  }
}

export async function isPoseidonAvailable(): Promise<boolean> {
  try {
    await loadPoseidon();
    return true;
  } catch {
    return false;
  }
}

export async function createPoseidonHashFn(): Promise<HashFn> {
  const poseidon = await loadPoseidon();
  const F = poseidon.F ?? { toObject: (x: any) => BigInt(x) };
  return (l: bigint, r: bigint) => BigInt(F.toObject(poseidon([l, r])));
}

// Default to sha256-based hash (not circuit-compatible) to avoid throwing in sync contexts.
export function defaultPoseidonHashFn(): HashFn {
  return (l: bigint, r: bigint) => sha256HashBigint(l, r);
}

export async function poseidonHash(...inputs: Array<bigint | number | string>): Promise<bigint> {
  const poseidon = await loadPoseidon();
  const F = poseidon.F ?? { toObject: (x: any) => BigInt(x) };
  const coerced = inputs.map((x) => BigInt(x));
  return BigInt(F.toObject(poseidon(coerced)));
}

export async function generateCommitment(value: bigint, blinding: bigint): Promise<bigint> {
  return poseidonHash(value, blinding);
}

export async function generateNullifier(secret: bigint, context: bigint): Promise<bigint> {
  return poseidonHash(secret, context);
}

// Fallback non-Poseidon hash (sha256) for testing / non-circuit contexts
export function sha256HashBigint(left: bigint, right: bigint): bigint {
  const buf = Buffer.concat([toFixedBuffer(left, 32), toFixedBuffer(right, 32)]);
  const digest = createHash('sha256').update(buf).digest('hex');
  return BigInt(`0x${digest}`);
}

function toFixedBuffer(value: bigint, size: number): Buffer {
  const hex = value.toString(16).padStart(size * 2, '0');
  return Buffer.from(hex, 'hex');
}

File: sdk/rust/src/crypto.rs
use anyhow::{anyhow, Result};
use ark_bn254::Fr;
use ark_ff::{BigInteger, Field, PrimeField};
use blake2::{Blake2s256, Digest};
use std::str::FromStr;

#[derive(Debug, Clone)]
pub struct MerkleProof {
    pub path: Vec<Fr>,
    pub indices: Vec<u8>,
    pub root: Fr,
    pub leaf: Fr,
}

pub struct MerkleTree {
    leaves: Vec<Fr>,
    layers: Vec<Vec<Fr>>,
}

impl MerkleTree {
    pub fn new(leaves: Vec<Fr>) -> Self {
        let mut padded = leaves;
        // pad to power of two
        let size = padded.len().next_power_of_two();
        while padded.len() < size {
            padded.push(Fr::ZERO);
        }
        let mut layers = Vec::new();
        layers.push(padded.clone());
        let mut current = padded;
        while current.len() > 1 {
            let mut next = Vec::with_capacity(current.len() / 2);
            for i in (0..current.len()).step_by(2) {
                let left = current[i];
                let right = current.get(i + 1).copied().unwrap_or(Fr::ZERO);
                next.push(poseidon_hash(&[left, right]));
            }
            layers.push(next.clone());
            current = next;
        }
        Self { leaves: layers[0].clone(), layers }
    }

    pub fn root(&self) -> Fr {
        *self.layers.last().unwrap().first().unwrap()
    }

    pub fn proof(&self, index: usize) -> Result<MerkleProof> {
        if index >= self.leaves.len() {
            return Err(anyhow!("index out of range"));
        }
        let mut path = Vec::new();
        let mut indices = Vec::new();
        let mut idx = index;
        for level in 0..self.layers.len() - 1 {
            let is_right = (idx % 2) == 1;
            let sibling_idx = if is_right { idx - 1 } else { idx + 1 };
            let sibling = self.layers[level].get(sibling_idx).copied().unwrap_or(Fr::ZERO);
            path.push(sibling);
            indices.push(if is_right { 1 } else { 0 });
            idx /= 2;
        }
        Ok(MerkleProof {
            path,
            indices,
            root: self.root(),
            leaf: self.leaves[index],
        })
    }

    pub fn verify(leaf: Fr, proof: &MerkleProof) -> bool {
        let mut acc = leaf;
        for (i, sibling) in proof.path.iter().enumerate() {
            let is_right = proof.indices.get(i).copied().unwrap_or(0) == 1;
            acc = if is_right { hash_pair(*sibling, acc) } else { hash_pair(acc, *sibling) };
        }
        acc == proof.root
    }
}

/// Convert a big-endian hex string into Fr
pub fn fr_from_hex(hex_str: &str) -> Result<Fr> {
    let cleaned = hex_str.trim_start_matches("0x");
    let bytes = hex::decode(cleaned)?;
    Ok(Fr::from_be_bytes_mod_order(&bytes))
}

/// Convert decimal string to Fr
pub fn fr_from_dec(dec_str: &str) -> Result<Fr> {
    let f = Fr::from_str(dec_str).map_err(|e| anyhow!("failed to parse dec to Fr: {e:?}"))?;
    Ok(f)
}

fn hash_pair(left: Fr, right: Fr) -> Fr {
    poseidon_hash(&[left, right])
}

/// Placeholder Poseidon: use Blake2s to derive a field element. Not circuit-compatible.
/// TODO: replace with real Poseidon parameters once available.
pub fn poseidon_hash(inputs: &[Fr]) -> Fr {
    let mut hasher = Blake2s256::new();
    for f in inputs {
        hasher.update(f.into_bigint().to_bytes_be());
    }
    let digest = hasher.finalize();
    Fr::from_be_bytes_mod_order(&digest)
}

pub fn generate_commitment(value: Fr, blinding: Fr) -> Fr {
    poseidon_hash(&[value, blinding])
}

pub fn generate_nullifier(secret: Fr, context: Fr) -> Fr {
    poseidon_hash(&[secret, context])
}
